### ISSUE TITLE
Add methods to return DOI and/or data citation

### ISSUE THREAD

**Issue by andypbarrett** [nsidc/earthaccess#203]
To make good data citation practices easier to follow, it would be great to have a method or methods that return the DOI and data citation for a dataset.  This would allow citations to be printed in notebooks or exported to bibliographies.
```
>results.get_doi()
10.7265/jj4s-tq79
>results.get_citation(format="APA")
Walsh, J. E., W. L. Chapman, F. Fetterer, and J. S. Stewart. (2019). Gridded Monthly Sea Ice Extent and Concentration, 1850 Onward, Version 2 [Data Set]. Boulder, Colorado USA. National Snow and Ice Data Center. https://doi.org/10.7265/jj4s-tq79. Date Accessed 02-20-2023.
>results.get_citation(format="bibtex")
@misc{Walsh, J. E., W. L. Chapman, F. Fetterer, and J. S. Stewart._2019, title={Gridded Monthly Sea Ice Extent and Concentration, 1850 Onward, Version 2}, url={https://nsidc.org/data/G10010/versions/2}, DOI={10.7265/jj4s-tq79}, publisher={National Snow and Ice Data Center}, author={Walsh, J. E., W. L. Chapman, F. Fetterer, and J. S. Stewart.}, year={2019} }
```

**Comment by mfisher87** [comment]
I love the idea of making it easy for people to cite. I imagine we'd want to generate citations from individual granules or individual collections? E.g. `results.get_doi()`, even if we have a `Results` class (#860), wouldn't always map to a single DOI.

**Comment by Sherwin-14** [comment]
@mfisher87 I would love to collaborate on this one. I tried retrieving **DOI** and **Citations** for _collections_ (although we are yet to decide whether this should be for granules and collections or any one of them), but I do have a questions.

I made a request for the concept_id C2789815241-ENVIDAT. In response there is UMM dictionary that contains two elements within it, one named `DOI` which contains **DOI** for the _collection_ and the other one named `CollectionCitations` which contains **Citations**. Below I have attached the request as well as the UMM part of the response.

Request:

```curl
curl "https://cmr.earthdata.nasa.gov/search/collections.umm_json?concept_id=C2789815241-ENVIDAT" | jq .
``` 

Response:

```JSON
{
  "umm": {
   ....
    "CollectionCitations": [
      {
        "Creator": "Katrin Meusburger, Elisabeth Graf Pannatier, Marcus Schaub",
        "Editor": "Katrin Meusburger",
        "DataPresentationForm": "url",
        "OnlineResource": {
          "Linkage": "https://www.envidat.ch/dataset/envidat-lwf-34"
        },
        "Publisher": "EnviDat",
        "Title": "10-HS Pfynwald",
        "ReleaseDate": "2019-01-01T00:00:00.000Z",
        "Version": "2019",
        "ReleasePlace": "Birmensdorf, Switzerland"
      }
    ],
   ....
    "DOI": {
      "DOI": "doi:10.16904/envidat.lwf.34"
    },
  ....
    "RelatedUrls": [
      {
        "URLContentType": "VisualizationURL",
        "Type": "GET RELATED VISUALIZATION",
        "URL": "https://www.envidat.ch/envidat_thumbnail.png"
      },
      {
        "URLContentType": "PublicationURL",
        "Type": "VIEW RELATED INFORMATION",
        "Subtype": "GENERAL DOCUMENTATION",
        "URL": "https://www.envidat.ch/dataset/envidat-lwf-34"
      }
    ]
  }
}
```
This is the response we can use to get **DOI** and **Citation** for each _collection_. But for some _collections_ we don't have data in these elements. For example, the concept_id C1214600621-SCIOPS, doesn't have `CollectionCitations` in its response, it does have `DOI` in it's response but it tells you that there is no information available for this _collection_.

```JSON
 {
  ....
  "DOI": {
          "MissingReason": "Unknown",
          "Explanation": "It is unknown if this record has a DOI."
        },
  ....
        "RelatedUrls": [
          {
            "Description": "Data Center top-level access page for this resource",
            "URLContentType": "DistributionURL",
            "Type": "GET DATA",
            "URL": "http://www.aoncadis.org/dataset/id/5c0c8206-d496-11de-892a-00c0f03d5b7c.html"
          }
        ],
}
```
Now if we click on the URL mentioned in the `RelatedUrls` , then it will redirect us to page where the **DOI** for the collection is present. The page also contains information on **Citations** as well for the collection.

What should we do in such cases where the response doesn't hold the contents of our interest (**DOI** and **Citations**) directly, but rather contains a link where those can be found? Asking this because it might be possible that the same link doesn't reside within `RelatedUrls` if we want to show the link as an output. Do we then just display the contents inside `DOI` part of the output? Although, for these two examples it does reside within  `RelatedUrls`.

**Comment by mfisher87** [comment]
>What should we do in such cases where the response doesn't hold the contents of our interest (DOI and Citations) directly, but rather contains a link where those can be found? 

I don't think we should support "spelunking" through links to find (and maybe scrape) information. I think we should use the information available on the API, and if collection owners want us to support showing any information, it's their responsibility to populate it in CMR :)

**Comment by mfisher87** [comment]
I think implementing this might look like developing a citation class which is capable of adapting the information provided in the API response to various citation formats, e.g. bibtex, APA, ..? as methods.

**Comment by jhkennedy** [comment]
We probably don't need to write our own code -- DOI foundation already provides a citation formatter:
https://citation.doi.org/

And it has a programmatic API: https://citation.doi.org/api-docs.html

But I like the idea of a requests wrapper that would call out to that API

**Comment by mfisher87** [comment]
Does a library underlie that HTTP API? I'd prefer to do such a simple computation locally instead of adding a new service dependency.

I think we'll need some bespoke logic for situations where a DOI is not present in the CMR response, but we can still generate a formatted citation.

**Comment by jhkennedy** [comment]
> Does a library underlie that HTTP API?

From a quick perusal, I don't see a python one that isn't just calling out to the API service under the hood:
https://pypi.org/search/?q=doi+citation

> I'd prefer to do such a simple computation locally instead of adding a new service dependency.

Generating a citation from a DOI is not a simple computation (my librarian wife is shouting this at me 🤣 ).

I have no interest in re-implementing that functionality when the DOI foundation itself is implementing a service for it, especially when it's just a GET request with 3 query parameters. But if you can find a library that does it, I'm open to it!

> I think we'll need some bespoke logic for situations where a DOI is not present in the CMR response, but we can still generate a formatted citation.

If there's no DOI and there is `CollectionCitations`, we could do this (I assume we'd prefer DOI?). I think if neither are present, we just don't, however.

**Comment by mfisher87** [comment]
> Generating a citation from a DOI is not a simple computation (my librarian wife is shouting this at me 🤣 ).

Perhaps not complexity-wise, but in terms of computational requirements, it's not something that should require a service, right?  The only reason we would use a service is because there is no local implementation we can leverage.

>If there's no DOI and there is CollectionCitations, we could do this (I assume we'd prefer DOI?). I think if neither are present, we just don't, however.

👍

**Comment by Sherwin-14** [comment]
> Perhaps not complexity-wise, but in terms of computational requirements, it's not something that should require a service, right? The only reason we would use a service is because there is no local implementation we can leverage.

My only qualm with local implementation approach would be that there will be some edge cases which we will need to figure out and handle. Instead of going through these on our own, we can rely on a service to take care of them for us. BTW,  should we get **DOI** and **Citations** for _collections_ only for now?

**Comment by mfisher87** [comment]
Yeah, to be clear, I _definitely_ don't think we should build our own local implementation :) I was hoping one already existed, and it seems it doesn't, so service it is!

**Comment by andypbarrett** [comment]
At least for NSIDC, DOI are only available at the collection level.
